### DISCUSSION TITLE
Discussion: Span of ref struct?

### DISCUSSION THREAD

**Discussion by HanabishiRecca** [dotnet/csharplang#2816] [Language Ideas]
Why we can't declare Span of ref struct?
```
ref struct A {...}
...
Span<A> arr = stackalloc A[10]; //error
```
Span itself is already a ref struct. So, if manually created ref struct of ref structs is ok, why we can't use Span of ref struct?
Of course ref struct can't be used as a generic type, but this is just a formal restriction. This means Span just needs some workaround.
E.g. such thing is possible in the unsafe context
```
var arr = stackalloc A[10];
```
We need something similar in the safe context.

**Comment by Joe4evr** [comment]
This would have to require something like #1148.

**Comment by quinmars** [comment]
> Of course ref struct can't be used as a generic type, but this is just a formal restriction. This means Span just needs some workaround.

It's more than that. How should the compiler know that you cannot pass your `Span<A>` to the following test function?

```csharp
public class C
{
    object remember;
    public bool Test<T>(Span<T> s)
    {
        remember = s.ToArray();
        return true;
    }
}
````

**Comment by HanabishiRecca** [comment]
@quinmars, exactly as I said.
In perfect case we need a native language feature for the stack array. (Also because Span is available only in the NET Core.)
Such as
```cs
A[] arr = stackalloc A[10];
var arr = stackalloc A[10]; //this will be interpreted as A[] - valid in the safe context
```
P.S. A[] is not the default array here, this is just an example, syntax may differ.

**Comment by yaakov-h** [comment]
> (Also because Span is available only in the NET Core.)

@HanabishiRecca Span is available on .NET Framework 4.5 and higher with a NuGet package, as well as Mono, Xamarin, UWP, and PCLs.
